CDS/js/youtube/youtube-upload-lib.js

```diff
@@ -0,0 +1,114 @@
+var Promise = require('promise');
+var fs = require('fs');
+var youtubeAuth = require('./youtube-auth.js');
+var googleapis = require('googleapis');
+var OAuth2 = googleapis.auth.OAuth2;
+var dataStore = require('../Datastore');
+
+const YOUTUBE_API_VERSION = 'v3';
+
+function getMetadata(connection) {
+
+    return new Promise((fulfill, reject) => {
+
+        if (!process.env.title) {
+            reject(new Error('Cannot upload to youtube: missing a title'));
+        }
+
+        if (!process.env.description) {
+            reject(new Error('Cannot upload to youtube: missing a description'));
+        }
+
+        if (!process.env.category_id) {
+            reject(new Error('Cannot upload to youtube: missing a category id'));
+        }
+
+        fulfill(dataStore.substituteStrings(connection, [process.env.title, process.env.description, process.env.category_id, process.env.access])
+        .then((substitutedStrings) => {
+            var title, description, category_id, status;
+            [title, description, category_id, status] = substitutedStrings;
+
+            return {
+                snippet: {
+                    title: title,
+                    description: description,
+                    categoryId: category_id
+                },
+                status: { privacyStatus: status ? status : 'private'}
+            }
+        }));
+    });
+}
+
+function getYoutubeData(connection) {
+    return this.getMetadata(connection)
+    .then((metadata) => {
+        const mediaPath = process.env.cnf_media_file;
+
+        if (!process.env.cnf_media_file) {
+            throw new Error('Cannot upload to youtube: missing media file path');
+        }
+
+        if (process.env.owner_channel) {
+            if (!process.env.owner_account) {
+                throw new Error('Cannot upload to youtube: missing account owner');
+            }
+        }
+
+        return dataStore.substituteStrings(connection, [process.env.cnf_media_file, process.env.owner_channel, process.env.owner_account])
+        .then((substitutedStrings) => {
+            var mediaPath, ownerChannel, ownerAccount;
+            [mediaPath, ownerChannel, ownerAccount] = substitutedStrings;
+
+            var youtubeData = {
+                'part': 'snippet,status',
+                'resource': metadata,
+                'media': {body: fs.readFileSync(mediaPath)},
```

Comment:
@Reettaphant @currysoup @clloyd I'm not sure whether doing a blocking read on the whole media file (could be into the hundreds of Mb, assuming it's compressed) is a good idea here, but not skilled enough with Javascript to be able to say with certainty.  What do you think?